Fix subclassing DatasetSchema

richard-rogers opened the pull request:
## Description

Make subclasses pay attention to changed default member values.

## Changes

- Make resolvers, type_mapper, default_config use default_factory so their defaults can be overridden in subclasses

## Related

richard-rogers committed:
Fix DatasetSchema subclassing

- Initialize resolvers member with fields()

richard-rogers committed:
- Make default_configs and type_mapper use default_factory

FelipeAdachi reviewed (commented):
I tested on my previous example, and it's still getting the resolver from the base class.

For reproducibility, the code snippet is:

```python
from whylogs.core.resolvers import Resolver
from whylogs.core.metrics import StandardMetric
from whylogs.core.schema import ColumnSchema, DatasetSchema
from dataclasses import dataclass, field
from whylogs.core.datatypes import DataType, Fractional, Integral
from typing import Dict, List
from whylogs.core.metrics.metrics import Metric



class MyCustomResolver(Resolver):
    """Resolver that keeps distribution metrics for Fractional and frequent items for Integral, and counters and types metrics for all data types."""

    def resolve(self, name: str, why_type: DataType, column_schema) -> Dict[str, Metric]:
        metrics: List[StandardMetric] = [StandardMetric.counts, StandardMetric.types]
        if isinstance(why_type, Fractional):
            metrics.append(StandardMetric.distribution)
        if isinstance(why_type, Integral):
            metrics.append(StandardMetric.frequent_items)


        result: Dict[str, Metric] = {}
        for m in metrics:
            result[m.name] = m.zero(column_schema.cfg)
        return result


@dataclass
class MyCustomSchema(DatasetSchema):
    resolvers = MyCustomResolver()


schema = MyCustomSchema()
print(schema.resolvers)
```

TheMellyBee commented on python/whylogs/core/schema.py:
```diff
@@ -56,9 +56,9 @@ class DatasetSchema:
-    default_configs: MetricConfig = MetricConfig()
-    type_mapper: TypeMapper = StandardTypeMapper()
-    resolvers: Resolver = StandardResolver()
+    default_configs: MetricConfig = field(default_factory=MetricConfig)
```
oof, this doesn't seem very readable especially as we make more custom ones like @FelipeAdachi example. Could this be iterated on?

TheMellyBee reviewed (changes requested):
Personally this doesn't seem very readable and easy to miss like Felipe did in even after the update. I think we could make this easy for users somehow

richard-rogers committed:
- Fix example DatasetSchema subclass
- Add custom resolver to example
- Fix DatasetSchema::copy() to work with subclasses

richard-rogers committed:
pre-commit
